Replace Checkboxes List with Stream Pills in Invite Users Modal #26871

sanchi-t opened the pull request:
<!-- Describe your pull request here.-->
Fixes #26871.

The Invite Users Modal previously featured a list of checkboxes for selecting streams when inviting users. This has been replaced with an input field featuring typeahead and stream pills for streamlined stream selection.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
## Before:
![image](https://github.com/zulip/zulip/assets/98596642/bd9169f4-a07a-4c11-a7e7-50fb36b178f9)

## After:

### Dark mode:

![image](https://github.com/zulip/zulip/assets/98596642/fc24a2ca-513b-4844-9a9e-aa2b2956c7eb)
![image](https://github.com/zulip/zulip/assets/98596642/11a85d0d-ade0-4abe-b9aa-66256a98b7d5)
![image](https://github.com/zulip/zulip/assets/98596642/716a11d7-ffca-4d70-ac5a-953e1dcfa681)


### Light Mode:

![image](https://github.com/zulip/zulip/assets/98596642/c089b261-6d8a-409c-93b0-5371aec8efb1)
![image](https://github.com/zulip/zulip/assets/98596642/9bb3e313-a84f-4937-adc7-70f9795c6faf)
![image](https://github.com/zulip/zulip/assets/98596642/6e26c4ce-1339-4832-9ac7-74153622b251)


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sanchi-t commented on web/src/stream_pill.ts:
```diff
@@ -70,10 +70,14 @@ export function get_user_ids(pill_widget: StreamPillWidget): number[] {
 }
 
-export function append_stream(stream: StreamSubscription, pill_widget: StreamPillWidget): void {
+export function append_stream(
+    stream: StreamSubscription,
+    pill_widget: StreamPillWidget,
+    display_value: (stream: StreamSubscription) => string,
+): void {
```
Is using a parameter, `show_stream_sub_size`, a better approach than passing a function?

The `show_stream_sub_size` parameter approach has been employed in a similar PR, #24580.

sanchi-t commented on web/src/add_streams_pill.ts:
```diff
@@ -0,0 +1,89 @@
+import * as input_pill from "./input_pill";
+import {set_up} from "./pill_typeahead";
```
Same dependency problem is faced by #24580. We probably have to wait for `pill_typeahead` ts conversion.

related CZO discussion [here](https://chat.zulip.org/#narrow/stream/6-frontend/topic/input_pill.20typescript/near/1670357).

timabbott replied:
It's probably reasonable to wait for that module to migrate, since it's being actively worked on.

sanchi-t commented on tools/test-js-with-node:
```diff
@@ -47,6 +47,7 @@ EXEMPT_FILES = make_set(
         "web/shared/src/poll_data.ts",
         "web/src/about_zulip.ts",
         "web/src/add_stream_options_popover.js",
+        "web/src/add_streams_pill.ts",
```
Should I write a test file for this instead?

timabbott replied:
It's not necessary unless the file is easy to write tests for, which this one may not be -- the goal is to document what we don't have tests for, not test everything.

timabbott commented:
I guess this is still blocked on migrating `pill_typeahead.js` to TypeScript? I think there's not much left to do that, as detailed in #frontend.

timabbott commented:
@sanchi-t the pill TypeScript work finally got integrated, so it should be possible to rebase and merge this one now!

shubham-padia commented on web/src/pill_typeahead.ts:
```diff
@@ -24,6 +25,18 @@ function group_matcher(query: string, item: UserGroupPillData): boolean {
+
+    return include_users || include_user_groups ? stream_query : () => stream;
+}
+
```
This function just returns a function, and is only used at one place. It might better to just define the function directly when assigning `include_streams`

sanchi-t replied:
Fixed.

shubham-padia commented on web/src/pill_typeahead.ts:
```diff
@@ -36,25 +49,42 @@ export function set_up(
-        opts.stream !== undefined && query.trim().startsWith("#");
+    if (!opts.help_on_empty_strings) {
+        opts.help_on_empty_strings = false;
+    }
```
Maybe we can use the ternary operator here.

sanchi-t replied:
Done.

shubham-padia commented on web/src/add_streams_pill.ts:
```diff
@@ -0,0 +1,78 @@
+
+    if (current_items.some((item) => item.type === "stream" && item.stream_id === sub.stream_id)) {
+        return undefined;
+    }
```
Why do we need this block of code? L27-L29

sanchi-t replied:
This code is almost same as `create_item_from_stream_name` function in `stream_pill.ts`. The L27-L29 ensures that duplicate stream pills are not added.

sanchi-t replied:
Yeah fixed it.

shubham-padia commented on web/templates/invite_user_modal.hbs:
```diff
@@ -67,33 +67,20 @@
             {{#if show_select_default_streams_option}}
             <div class="select_default_streams new-style">
-                <label class="checkbox display-block">
+                <label class="checkbox display-block modal-field-label">
```
Why have we added the `modal-field-label` class?

sanchi-t replied:
To maintain consistent margins between the labels and input fields in the modal, and as suggested by this CZO [comment](https://chat.zulip.org/#narrow/stream/101-design/topic/User.20invite.3B.20streams.20they.20should.20join/near/1724788).

shubham-padia reviewed (commented):
Looks good visually, thanks @sanchi-t ! I've left a few comments, it also might be worth investing some time into making the commit structure of this PR a bit better

sanchi-t commented on web/src/stream_pill.ts:
```diff
@@ -64,10 +64,11 @@ export function get_user_ids(pill_widget: StreamPillWidget | CombinedPillContain
 export function append_stream(
     stream: StreamSubscription,
     pill_widget: CombinedPillContainer,
+    display_value: (stream: StreamSubscription) => string,
```
Fixed it.

shubham-padia reviewed (changes requested):
Left some comments. Moreover, the last 3 commits should be squashed since the css and test changes of a feature should be in the same commit.

sanchi-t commented on web/src/pill_typeahead.ts:
```diff
@@ -36,25 +37,42 @@ export function set_up(
+        opts.stream && query.trim().startsWith("#");
+
+    const include_streams = include_users || include_user_groups ? stream_query : () => opts.stream;
+
```
Done.

sanchi-t commented on web/src/pill_typeahead.ts:
```diff
@@ -36,25 +37,42 @@ export function set_up(
+        display_pill = stream_pill.display_pill;
+    } else {
+        display_pill = opts.display_pill;
+    }
```
Done.

timabbott replied:
I'm not entirely sure this version is easier to read; probably the `??` operator might be clearest?

timabbott replied:
So also of note is that this function being patched here is a generic function not limited to just streams, and yet we have a function called `display_pill` that is only relevant for those? That doesn't make sense.

timabbott replied:
Do we need a default value at all?

sanchi-t replied:
> Do we need a default value at all?

No, we use `display_pill` when we call `stream_pill.append`. We can change the `stream_pill.append` function such that it accepts an optional parameter for getting stream pill label and it would provide the default value.

sanchi-t replied:
> So also of note is that this function being patched here is a generic function not limited to just streams, and yet we have a function called display_pill that is only relevant for those? That doesn't make sense.

We can rename the function so it indicates its stream specific like `stream_display_pill`?

I think we need this function in `pill_typeahead.ts` because `pill_typeahead` calls `stream_pill.append` and we need to mention the display value of pill there. It wouldn't been have a problem if we displayed the same format value for stream pill everywhere, but for user invite modal we just want stream name as the display value and not extra information about subscriber count.

 Also you suggested to have `stream_match_prefix` as a parameter which is also specific to streams so i think a function specific to streams should also be alright.

sanchi-t commented on web/src/stream_pill.ts:
```diff
@@ -22,6 +22,7 @@ export function display_pill(sub: StreamSubscription): string {
 export function create_item_from_stream_name(
     stream_name: string,
     current_items: CombinedPillItem[],
+    display_pill_func?: (sub: StreamSubscription) => string,
```
not sure about the parameter name.

sanchi-t replied:
Because of this line:
```
display_value: display_pill_func ? display_pill_func(sub) : display_pill(sub),
````
If we change the parameter name to `display_pill` then the code will always refer to the parameter even when we want to call the `stream_pill.display_pill`.

shubham-padia replied:
Ah, the name seems fine in that case.

timabbott replied:
My feedback on naming elsewhere applies, but one extra note is that we don't use abbreviations like "_func" as a project policy. A good name shouldn't require such a suffix anyway; something like `stream_to_pill_label` is going to be much better in most cases anyway.

sanchi-t replied:
Okay, I noticed `update_func` in `opts` within `pill_typeahead.ts` and thought we could utilize that suffix. However, that seems to be a rare use case for suffixes, I guess.

shubham-padia reviewed (approved):
LGTM, this is ready for integration review

timabbott commented on web/src/stream_pill.ts:
```diff
@@ -14,7 +14,7 @@ type StreamPillWidget = InputPillContainer<StreamPill>;
 export type StreamPillData = StreamSubscription & {type: "stream"};
 
-function display_pill(sub: StreamSubscription): string {
+export function display_pill(sub: StreamSubscription): string {
```
This should probably be renamed to `format_stream_name_and_subscriber_count`, and the parameter renamed to from `display_value` to `stream_to_display_value`, to make it more obviously a function.

Also of note is that this (existing) function is buggy -- it doesn't do i18n and thus returns things always in English. See
https://zulip.readthedocs.io/en/latest/translating/internationalization.html

sanchi-t replied:
Check this comment out please [here](https://github.com/zulip/zulip/pull/28686#discussion_r1635255688).

timabbott commented on web/src/pill_typeahead.ts:
```diff
@@ -36,25 +37,40 @@ export function set_up(
+    const include_streams: (query: string) => boolean | undefined = (query) => {
+        if (include_users || include_user_groups) {
+            return opts.stream && query.trim().startsWith("#");
+        }
```
This might be better written by just having an explicit `stream_match_prefix` parameter that would be `#` for the callers that want such a prefix, which happens to be the ones that are only including streams? Seems like a more orthogonal/correct API.

sanchi-t replied:
> This might be better written by just having an explicit stream_match_prefix parameter that would be # for the callers that want such a prefix, which happens to be the ones that are only including streams?

Callers which only include streams should have the `stream_match_prefix` value a s "" (empty string). And for the callers which include streams and users/user groups should have the `stream_match_prefix` value as "#".

timabbott replied:
I am not sure I understand, but yeah I think that would be fine.

timabbott commented on web/src/invite.ts:
```diff
@@ -358,6 +357,10 @@ function open_invite_user_modal(e: JQuery.ClickEvent<Document, undefined>): void
 
+        const $stream_pill_container = $("#invite_streams_container .pill-container");
+
+        stream_pill_widget = add_streams_pill.create($stream_pill_container);
```
So `render_invite_user_modal` is passed a `streams` parameter, and maybe some others, that are now unused; probably needs various bits of code deleted. Notably that started with the set `get_invite_streams(); I'm not sure your new `add_streams_pill` module uses the correct logic?

sanchi-t replied:
> I'm not sure your new `add_streams_pill` module uses the correct logic?

The function `get_invite_streams()` currently returns all default streams plus the streams the inviting user is subscribed to, sorted alphabetically. However, since we're now using pills, alphabetical order is unnecessary because users can search for streams directly. Additionally, do you want to restrict the search results in the input box? Currently, users can search through all streams, do you want to limit it to only default streams and the streams the inviting user is subscribed to?

timabbott replied:
Hmm, yeah. I think we want to limit to exactly the set of streams that the current user as permission to add another user to. 

But also if you're removing the last caller to `get_invite_streams`, we probably want to delete that function. But it might be just as reasonable to rework the function to use the correct rule, removing the sorting, and still call it.

sanchi-t replied:
I suggest removing the `get_invite_stream`s function. Its functionality is essentially redundant as it merely sorts the results provided by `stream_data.get_invite_stream_data`, which already performs the necessary work.

We should use and rework `stream_data.get_invite_stream_data` instead.

timabbott commented on web/src/add_streams_pill.ts:
```diff
@@ -0,0 +1,64 @@
+    add_default_stream_pills(pill_widget);
+    set_up_pill_typeahead({pill_widget, $pill_container: $stream_pill_container});
+    return pill_widget;
+}
```
How did you pick the name for this module? I think the name is too generic, since it's not a generic streams picker, but instead one specifically designed for the invitations stream picker; a good name would reflect that, like `invite_stream_picker_pill.ts`

sanchi-t replied:
Yeah I am pretty bad at naming files and functions.

timabbott replied:
It's a hard skill! I think the key to improving is to try spend a few minutes thinking about each larger thing (function/module/etc.) you name, and then you can check your reasoning by including a comment in your PRs next to each name you introduce explaining how you picked it.

sanchi-t replied:
Thank you, I'll try this.

timabbott commented:
@sanchi-t are you able to do the finishing work on this in the next few days? If not, it may make sense for @shubham-padia to do the revisions just so we can get it integrated while our reviews are still fresh.

sanchi-t commented:
@timabbott Yes but I have some more questions and think the PR will require more rounds of review.

sanchi-t commented on web/src/stream_pill.ts:
```diff
@@ -16,7 +17,10 @@ export type StreamPillData = StreamSubscription & {type: "stream"};
 function display_pill(sub: StreamSubscription): string {
     const sub_count = peer_data.get_subscriber_count(sub.stream_id);
-    return "#" + sub.name + ": " + sub_count + " users";
+    return $t(
+        {defaultMessage: "#{stream_name}: {sub_count} users"},
+        {stream_name: sub.name, sub_count},
+    );
```
I don't think this is right. The whole display value like "#stream: 15 users" can't be translated. Only "users" can be translated here. I think we don't translate numbers for now. To make this "#stream: 15 users" translation work, we'd have to create many combinations in the locale file with different stream names and numbers, which isn't practical. I think it should be something like:

```
function format_stream_name_and_subscriber_count(sub: StreamSubscription): string {
    const sub_count = peer_data.get_subscriber_count(sub.stream_id);
    return "#" + sub.name + ": " + sub_count + " " + $t({defaultMessage: "users"}, {sub_count},
    );
}
```

But this method isn't correct because we're adding a string with `${}` elements. To make translation work properly, I think we need to change the `InputPillItem` and modify the `display_value`. Instead of being just a string, `display_value` should be made up of a complex structure. This structure would allow us to separate translatable and untranslatable parts of the sentence.  However, I'm not sure if it's worth making such big and complicated changes. We should try to find a way to write some specific code for pills that can ignore certain parts when we translate them.

sanchi-t replied:
Update: Separated "users" from the display value so that it could be translated properly.

sanchi-t commented on web/src/stream_data.ts:
```diff
@@ -365,29 +365,19 @@ export function get_streams_for_user(user_id: number): {
-        };
-    }
-
+export function get_invite_stream_data(): StreamSubscription[] {
```
Should this function be modified to also include all the public streams?

timabbott replied:
I think you should try posing the question as to what needs to be decided in #frontend, with appropriate context. Probably someone will be able to help.

sanchi-t replied:
Here is the [CZO ](https://chat.zulip.org/#narrow/stream/6-frontend/topic/invite.20streams/near/1824320) chat.

timabbott commented:
@sanchi-t can you comment on the state of this? I saw you did a push, but am not sure if it's just merge conflicts or attempts to resolve the remaining feedback. I'd really like to merge this during the current week for it to be on track for 9.0.

sahil839 commented:
Regarding the approach for implementing pills, I think we can have a separate functions for stream only pills, like the #30344 and #24580. And I guess that would also help us in passing the `stream_matching_prefix` as you do not need to check prefix if only stream pills are allowed.

sanchi-t commented:
@timabbott, I've taken care of all the feedback except for Sahil's latest comment. I plan to address that within a day. I had some issues and questions that I have expressed in comments.

sanchi-t commented on web/src/pill_typeahead.ts:
```diff
@@ -79,17 +80,27 @@ export function set_up_combined(
         exclude_bots?: boolean;
         update_func?: () => void;
+        help_on_empty_strings?: boolean;
+        invite_streams?: boolean;
```
Update: Created `set_up_stream` instead if using `set_up_combined`.

sanchi-t commented:
Updates since last review:
 -  Following Sahil's suggestion, I created a new function called `set_up_streams` instead of using `set_up_combined`.
 -  Added properties to the `StreamPill` type to handle the stream name and subscriber count separately. This allows for proper translation in handlebars, specifically separating "users" from the display format like `#stream_name: 5 users`. We do not translate the stream name or number of subscribers.
-  Ensured that only streams where the user has permission to add others appear in the typeahead results.

NOTE:

- I am creating a new `create_item_from_stream_name` function now instead of using one in `stream_pill` that was suggested by Shubham since the function (`invite_stream_picker_pill.create_item_from_stream_name`) is bit different now ([here](https://github.com/zulip/zulip/pull/28686#discussion_r1623723481)).
- The renaming of the function from `display_pill` to `format_stream_name_and_subscriber_count` is not valid anymore. I split that function into two separate functions to fix a translation bug ([here](https://github.com/zulip/zulip/pull/28686#discussion_r1630490219)).

sahil839 commented on web/src/invite_stream_picker_pill.ts:
```diff
@@ -0,0 +1,63 @@
+
+function create_item_from_stream_name(
+    stream_name: string,
+    current_items: CombinedPillItem[],
```
I think this can still reuse the `stream_pill` function by passing a parameter like `stream_prefix_required` and `show_stream_sub_count`.

sanchi-t replied:
Done.

sahil839 commented on web/src/pill_typeahead.ts:
```diff
@@ -72,6 +72,52 @@ export function set_up_user(
+        matcher(item: StreamPillData, query: string): boolean {
+            query = query.toLowerCase();
+            query = query.replaceAll("\u00A0", " ");
+            query = query.trim().slice(1);
```
I don't think `slice` is required here since we do not have `#` prefix in this case.

sahil839 replied:
Or maybe we can just also match the streams if `#` is present, so we should only slice if first character is `#`.

sanchi-t replied:
Fixed.

sahil839 commented on web/src/pill_typeahead.ts:
```diff
@@ -72,6 +72,52 @@ export function set_up_user(
+                assert(match.type === "stream");
+                stream_matches.push(match);
+            }
+            return typeahead_helper.sort_streams(stream_matches, query.trim().slice(1));
```
Similar to above, `slice` should not be needed always here.

sahil839 commented on web/src/stream_data.ts:
```diff
@@ -365,32 +365,10 @@ export function get_streams_for_user(user_id: number): {
-
+export function get_invite_stream_data(): StreamSubscription[] {
+    const user_id = people.my_current_user_id();
+    const {subscribed, can_subscribe} = get_streams_for_user(user_id);
```
I think this will work because we would not call this code if user is not allowed to subscribe others as the widget itself will not be visible, but the logic used to `can_subscribe` considers whether the user can subscribe himself or not and not subscribe others.

So, should be better to just fitler the streams by using `can_subscribe_others` function so that we do not accidentally break it while changing how the permission to subscribe others is computed.

sahil839 reviewed (commented):
Posted some comments.

Also, the invite modal is not opened if the user does not have permission to susbcribe others, so that should be fixed.

sahil839 commented:
Also, I guess the previous approach that you used for translating the text inside the pills [here](https://github.com/zulip/zulip/pull/28686#discussion_r1635255688 was fine since variable values are not translated. You can see we use similar methods when text contains user on stream name, one is where we translate the "Already subscribed to {channel}" string in `stream_settings_components.js`.

sanchi-t commented on web/src/pill_typeahead.ts:
```diff
@@ -79,17 +79,24 @@ export function set_up_combined(
         user_source?: () => User[];
         exclude_bots?: boolean;
         update_func?: () => void;
+        help_on_empty_strings?: boolean;
```
Yeah, I am not sure. I thought Tim wanted that `stream_match_prefix` property for `opts` in `set_up_combined`.
Anyway I removed that commit.

sanchi-t replied:
Fixed.

sanchi-t commented on web/src/stream_pill.ts:
```diff
@@ -8,15 +8,20 @@ export type StreamPill = {
+}
+
+function stream_subscriber_count(sub: StreamSubscription): number {
+    return peer_data.get_subscriber_count(sub.stream_id);
```
Fixed.

sanchi-t commented:
> Also, the invite modal is not opened if the user does not have permission to susbcribe others, so that should be fixed.

What do you mean? If user does not have permission to subscribe other the invite modal should not be opened and that is the current behavior. I am not getting what you mean.

sanchi-t commented:
> Also, I guess the previous approach that you used for translating the text inside the pills [here](https://github.com/zulip/zulip/pull/28686#discussion_r1635255688 was fine since variable values are not translated. You can see we use similar methods when text contains user on stream name, one is where we translate the "Already subscribed to {channel}" string in stream_settings_components.js.

So what do you suggest? Use the previous approach? I think the current approach is better as it is easier/simpler to use `stream_pill.create_item_from_stream_name`.

sahil839 commented:
> > Also, the invite modal is not opened if the user does not have permission to susbcribe others, so that should be fixed.
> 
> What do you mean? If user does not have permission to subscribe other the invite modal should not be opened and that is the current behavior. I am not getting what you mean.

It should open and the current behavior that it does not open.

sahil839 commented on web/src/stream_pill.ts:
```diff
@@ -12,29 +12,41 @@ export type StreamPill = {
         return undefined;
     }
 
+    if (invite_streams) {
```
One approach here could be to pass a allowed streams list here. Or some validation function like `can_subscribe_others` here. This would mean that that could be used in future for cases other than invite UI.

sanchi-t replied:
Yeah I passed a validation function as param.

sahil839 commented on web/tests/stream_data.test.js:
```diff
@@ -944,7 +947,7 @@ test("get_invite_stream_data", () => {
         stream_id: 25,
         invite_only: true,
-        default_stream: false,
+        subscribed: true,
```
Since we are updating the test here, we can test the complete logic like unsubscribed private streams are not included.

sanchi-t replied:
Done.

sahil839 commented on web/src/stream_pill.ts:
```diff
@@ -12,29 +12,41 @@ export type StreamPill = {
 export type StreamPillData = StreamSubscription & {type: "stream"};
 
-function stream_display_pill(sub: StreamSubscription): string {
+export function stream_display_pill(sub: StreamSubscription): string {
```
I think this change is not needed anymore.

sanchi-t replied:
Yeah right.

sahil839 commented:
> So what do you suggest? Use the previous approach? I think the current approach is better as it is easier/simpler to use stream_pill.create_item_from_stream_name

I am not sure. `create_item_from_stream_name` already accepts `show_subscriber_count` parameter, so should be easy to just compute the string there itself. It will also help in avoiding adding two extra parameters to `StreamPill` and other types.

sanchi-t commented:
Updates since last review:
- Used `$t` for translation (previous approach) instead of adding new properties to `StreamPill`.
- The invite modal now opens if user does not have permission to subscribe other.
- Addressed the review comments.

timabbott commented on web/tests/pill_typeahead.test.js:
```diff
@@ -224,6 +224,99 @@ run_test("set_up_user", ({mock_template, override, override_rewire}) => {
+        assert.equal(typeof config.highlighter_html, "function");
+        assert.equal(typeof config.matcher, "function");
+        assert.equal(typeof config.sorter, "function");
+        assert.equal(typeof config.updater, "function");
```
I'm not really a fan of the testing strategy of doing so much mocking.

timabbott commented on web/src/stream_pill.ts:
```diff
@@ -67,33 +77,43 @@ export function get_user_ids(pill_widget: StreamPillWidget | CombinedPillContain
     stream: StreamSubscription,
-    pill_widget: CombinedPillContainer,
+    pill_widget: StreamPillWidget | CombinedPillContainer,
+    show_subscriber_count = true,
```
It would be better to do the refactors to add options like this in their own commits, for future reference, since that helps a reviewer model how much risk there is of breaking other pill typeahead components.

timabbott commented:
So I did a fairly extensive review, and I think this is basically good to merge. I ended up squashing the last two commits together, since that seemed more readable to me; the new final commit essentially fixed a correctness issue in the previous one. I found one bug, which I'll open as a follow-up issue that should be a high priority -- I don't intend to deploy this to production until it's fixed:

- After you `backspace` to remove a stream pill, the pill typeahead opens, preventing you from using `backspace` to remove a second stream pill.
